tools/gen-doctest.py

```diff
@@ -0,0 +1,182 @@
+
+    def process(self, infile):
+        """
+        Find doctests in a Markdown file and process them according to the
+        constructor parameters.
+
+        :param infile: path to the input file.
+        """
+        self._infile = infile
+        self._outname_base = os.path.splitext(os.path.basename(infile))[0]
+        self._outname_cnt = 1
+
+        mode = 'TEXT'
+        decl, decl_lineno = {}, 0
+        code, code_lineno = [], 0
+
+        for line in fileinput.input(infile):
+            decl_match = re.match(r'^\[doctest\]:\s+#\s+\((.*)\)\s*$', line)
+            nl_match = re.match(r'^\s*$', line)
+            start_match = re.match(r'^```c\s*$', line)
+            end_match = re.match(r'^```\s*', line)
+
+            if mode == 'TEXT':
+                if decl_match is not None:
+                    decl, decl_lineno = self._process_decl(decl_match.group(1))
+                    mode = 'NL'
+            elif mode == 'NL':
+                if decl_match is not None:
+                    self._warning('test without code block', decl_lineno)
+                    decl, decl_lineno = self._process_decl(decl_match.group(1))
+                elif start_match is not None:
+                    code, code_lineno = self._process_code_start()
+                    mode = 'CODE'
+                elif nl_match is None:
+                    self._warning('test without code block', decl_lineno)
+                    mode = 'TEXT'
+            elif mode == 'CODE':
+                if end_match is not None:
+                    self._process_code_end(decl, code)
+                    mode = 'TEXT'
+                else:
+                    code.append(line)
+
+        if mode == 'NL':
+            self._warning('test without code block', decl_lineno)
+        elif mode == 'CODE':
+            self._warning('unterminated code block', code_lineno)
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Markdown doctest extractor', epilog="""
+        The tool extracts specially marked fenced C code blocks from the input Markdown files
+        and writes them to the file system. The annotations recognized by the tool are special
+        but valid Markdown links/comments that must be added before the fenced code blocks:
+        `[doctest]: # (name="test.c", ...)`. For now, two parameters are valid:
+        `name` determines the filename for the extracted code block (overriding the default
+        auto-numbered naming scheme), and `test` determines the test action to be performed on
+        the extracted code (valid options are "compile", "link", and the default "run").
+        """)
+    parser.add_argument('-d', '--dir', metavar='NAME', default=os.getcwd(),
+                        help='output directory name (default: %(default)s)')
+    parser.add_argument('--dry', action='store_true',
+                        help='don\'t generate files but print file names that would be generated '
```

Comment:
Yes, it's all about personal preferences. (I made my choice because the repr of strings is delimited by single quotes in python by default, so I try to stick to them whenever possible.) So, if it's not strongly required, I'd keep this single escape here as is.